idle.go

```diff
@@ -0,0 +1,365 @@
+	// therefore we should not ask the ClientConn to enter idle mode.
+	if !atomic.CompareAndSwapInt32(&i.activeCallsCount, 0, math.MinInt32) {
+		return false
+	}
+	// It is possible that one or more RPCs started and finished since the last
+	// time we checked the calls count and activity in the timer callback. In
+	// this case, the calls count would be zero and the above compare-and-swap
+	// operation would have succeeded. The `activeSinceLastTimerCheck` field is
+	// set to false only from the timer callback and is set to true in
+	// onCallBegin. Therefore, if one or more RPCs started and finished before
+	// we got here, this field would be set to true. And in this case, we don't
+	// want to enter idle mode.
+	if active := atomic.LoadInt32(&i.activeSinceLastTimerCheck); active == 1 {
+		return false
+	}
+	if err := i.enforcer.enterIdleMode(); err != nil {
+		logger.Errorf("Failed to enter idle mode: %v", err)
+		return false
+	}
+	// We are in enter idle mode now. Set the active calls count appropriately.
+	atomic.StoreInt32(&i.activeCallsCount, math.MinInt32)
+	return true
+}
+
+// exitIdleMode instructs the ClientConn to exit idle mode.
+//
+// Holds idleMu which ensures mutual exclusion with enterIdleMode.
+func (i *atomicIdlenessManager) exitIdleMode() error {
+	i.idleMu.Lock()
+	defer i.idleMu.Unlock()
+
+	// When the channel enters idle mode, the active calls count is set to
+	// math.MinInt32. RPCs that start when the channel is in idle mode increment
+	// the active calls count in onCallBegin. If there are multiple RPCs
+	// competing to get the channel out of idle mode, the first one to grab the
+	// lock here gets to do so and sets the active calls count back to 1. So, if
+	// the count is not negative here, we know that someone else won the race
+	// and moved the channel out of idle mode, and we have nothing to do here.
+	if count := atomic.LoadInt32(&i.activeCallsCount); count >= 0 {
+		atomic.AddInt32(&i.activeCallsCount, 1)
+		return nil
+	}
+
+	// The first one to grab the lock will get here to instruct the channel to
+	// move out of idle mode.
+	if err := i.enforcer.exitIdleMode(); err != nil {
+		return status.Errorf(codes.Internal, "grpc: ClientConn failed to exit idle mode: %v", err)
+	}
+
+	// Reset the calls count to 1 and reset the timer to fire after a duration
+	// of the configured idle timeout.
+	atomic.StoreInt32(&i.activeCallsCount, 1)
+	i.timer.Store(timeAfterFunc(time.Duration(i.timeout), i.handleIdleTimeout))
+	return nil
+}
+
+// handleIdleTimeout is the timer callback that is invoked upon expiry of the
+// configured idle timeout. The channel is considered inactive if there are no
+// ongoing calls and no RPC activity since the last time the timer fired.
+func (i *atomicIdlenessManager) handleIdleTimeout() {
+	if i.isClosed() {
+		return
+	}
+
+	var timeoutDuration time.Duration
+	if count := atomic.LoadInt32(&i.activeCallsCount); count > 0 {
```

Comment:
I think this may be a slight improvement(?).  There should be fewer checks anyway:

```go
func handleIdleTimeout() {
  if atomic.LoadInt32(&i.activeSinceLastTimerCheck) > 0 {
    // there's been activity.  set timer & return
  }

  if atomic.CompareAndSwapInt32(&i.activeCallsCount, 0, -math.MaxInt32 /* hmmm see below */) {
    // extremely likely to succeed given no activity since last timer.
    // timer fired, no RPCs are in flight, or happened since last timer.
    // blocking new RPCs briefly while we attempt to enter idle.
    if tryEnterIdleMode() {
      // entered idle mode; no timer needed; we are done
      return
    }
    // failed to enter idle mode due to a concurrent RPC that "exited idle mode", or error from channel.
    atomic.AddInt32(&i.activeCallsCount, math.MaxInt32) // undo attempt to enter idle
  }
  // set timer and return
}

func tryEnterIdleMode() bool {
  i.mu.Lock()
  defer i.mu.Unlock()
  if atomic.LoadInt32(&i.activeCallsCount) == -math.MaxInt32 {
    // no new RPCs have come in and we have the lock. enter idle if we can.
    i.enforcer.enterIdleMode() // handle err
    // on success:
    i.actuallyIdle = true
    return true
  }
  // oops we raced and lost to a new RPC. very rare, but stop entering idle.
  return false
}

func onCallBegin() error {
  if atomic.AddInt32(&activeCallsCount, 1) > 0 {
    // not idle. allow call.
    atomic.StoreInt32(&i.activeSinceLastTimerCheck, 1)
    return nil
  }
  // activeCalls was negative; exit idle to allow this RPC
  i.exitIdleMode() // fixes activeCalls atomic.  handle error
  atomic.StoreInt32(&i.activeSinceLastTimerCheck, 1)
  return err
}

func exitIdleMode() error {
  i.mu.Lock()
  defer i.mu.Unlock()

  if !i.actuallyIdle {
    // we were about to go idle but this RPC just barely won the race for taking the lock. very rare.
    // we won't go idle in this case because of the activeCallsCount inc which will prevent tryEnterIdleMode
    //
    // or, multiple new RPCs were all attempting to exit idle together and an earlier one took us out of idle.
    // either way, nothing to do here.
    return nil
  }

  i.enforcer.exitIdleMode() // handle error
  // undoes idle entry process, respecting any new RPC attempts
  atomic.AddInt32(&i.activeCallsCount, math.MaxInt32)
  i.actuallyIdle = false
  // start new timer to check for idleness
}
```

Tricky interactions here, but I think this works.

Reply:
Suggestion applied. Thanks.